### ISSUE TITLE
Let's make the lab guides look great again!

### ISSUE THREAD

**Issue by goetzrieger** [ansible/workshops#838]
##### SUMMARY
The way we present the lab guides right now is not too attractive and error-prone in parts. Here is a way to, well, make them look better! ;)

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
- exercises
- docs

##### ADDITIONAL INFORMATION
The lab guides for the Linux Automation labs where initially authored in asciidoc(tor) by me. I understand why the move to Markdown was necessary (github pages) and appreciate the work that went into it. But: In my opinion the HTML rendered by asciidoctor looks a lot clearer, more structured, nicer admonitions, numbering, TOCs and on. All in all a lot more user friendly.

Me and team run six Ansible lab sessions at virtual Summit. We wanted to have something more eye-pleasing in terms of lab guides and @cbolz came up with Hugo and a really nice theme. Hugo renders static html from markdown that can easily be hosted all over the place.  

Here are our lab guides: https://goetzrieger.github.io/

Sources are here: https://github.com/goetzrieger/ansible-labs-summit2020

I believe this really is something we should discuss!

**Comment by IPvSean** [comment]
that is pretty, i want it

one of the other ideas we have been discussing is using patternfly which is used for cloud.redhat.com, redhat.com and ansible tower.  They have GitHub pages setup here-> https://github.com/patternfly/patternfly.github.io

which looks like this-> https://patternfly.github.io/v4/

**Comment by goetzrieger** [comment]
Patternfly looks pretty cool, sure. But I was under the impression it's mainly for applications. I'm a bit worried that there would be too much overhead involved for simply publishing docs...

**Comment by cbolz** [comment]
having a quick look at the patternfly documentation, there seems to be nothing on how to use this in Jekyll or any other site generator. Unless you want to write the theme from scratch, this sounds like quite some work.

**Comment by IPvSean** [comment]
You can see they used GitHub pages to generate that page though. I linked
their assets repo

**Comment by jgroom33** [comment]
I suggest using http://diagrams.net/ to make diagrams.

one of following workflows works well:

1. export as png (check the box "include a copy of my diagram")
   a. with that checkbox it embeds the svg object so it can be updated
2. save as foo.drawio.png

The diagrams can then be both embedded and edited easily.

Aslo, use asciidoctor instead of markdown. It's easier to embed file contents.

**Comment by IPvSean** [comment]
i am not opposed to moving to ascii-doc but I figured this might help folks think a bit about options

- github pages supports ascii doc, so that is not a problem
- markdown was initially a requirement of lightbulb
- the "good looking" you linked has more to do with CSS then ascii-doc versus markdown, I think we can replicate the results pretty easily... I have been waiting on funding to hire a CSS developer but we can just do it
- we have the lab guides translated into several languages so if we convert to ascii-doc we need folks to volunteer to help shift the guides over, I am not sure ROI of that if we just fix the problem which is making them look good

**Comment by cigamit** [comment]
For my tokenized instant workshops I have been wrestling with this a bit more.  This is what I currently am doing.

I take the github markdown pages, convert them to html (using [this](https://github.com/KrauseFx/markdown-to-html-github-style) and a quick build script), add a custom style sheet, and then host them on the students tower servers as another url (student#-docs).  This allows me to customize the docs per student, which fixes a lot of the copy and paste errors we see (actual student number in urls and forms).

What I am starting to do now is make the images themselves "generated" per student.  So I am taking the images, changing them to be background images on divs that then contains a SVG object.  That SVG object is just the text I want to display overlaid on the image.  So now every image doesn't display student1, but instead displays exactly what they should be seeing when they type in their student number.  I may change the images to display all typed fields like this.  This will solve issues where we have to edit the images anytime we change the name of something (like the Windows Workshop Inventory).

With github pages this still works fine (it allows you to use direct html in them too), but its not as clean looking.

example code
```
<div style="width: 893px; height:343px; background-image: url('images/1-tower-add-machine-credential.png')">
<svg viewBox="0 0 893 343">
  <style>
    .small { font: 14px "Open Sans";}
  </style>
  <text x=54 y=321 class=small>student1234</text>
</svg>
</div>
```
resulting image
![image](https://user-images.githubusercontent.com/957322/97528905-c6fc6180-197c-11eb-8b39-7294cef8dac6.png)

**Comment by cloin** [comment]
I really like how this would be just as easy to collab on docs as it is today using markdown. How much time does per student documentation add to provisioning?

**Comment by cigamit** [comment]
Per student isn't difficult.  Granted, you can't do it the exact way I do it, as I am using my fast provisioner with them already baked into the host (and so I just need a quick sed command on the files to replace the student number).   I did all my building of the html files before provisioning, which is an option we can also do, we would just have to regenerate them after edits to the docs.

In essence you would just need to do this (assuming Windows Workshop, since that's the only one I ever look at)

1. Ensure dependencies are installed on provisioning host for script (will depend on which one we use)
2. Add another nginx conf to the Ansible Tower server for studentX-docs (just like we do with -code) with a root pointing to /opt/docs/  (Could be docs.student...)
3. Generate an SSL cert for it (I use wildcards certs and a proxy, so I don't have to do this step)
4. Generate docs using script
5a.  Copy the generated files to remote server and use a find / sed combo to replace student# with actual student
5b.  or use a loop and template module to do the replacement.  It means our current docs would before jinja2 templates (which them opens a lot of other possibilities)
6. Ensure all docs are using the same exact string for representing the student number (I actually use student####) since all my student numbers are 4 numbers.

All in all, it only adds a few seconds but depends on what hoops you need for setting up the script and whether you are looping the template module or not.  The one I used just needs npm installed, clone down the repo for it, and then "npm install" within that directory to get the dependencies.  I think there are some pure python scripts out there also that don't require all those hoops.

**Comment by IPvSean** [comment]
@goetzrieger we had a UX person come and update our workshop page to match ansible.com
https://ansible.github.io/workshops/

![image](https://user-images.githubusercontent.com/6537680/100159793-66c5e600-2e7c-11eb-9892-f396ae7a26d8.png)

for the "per student" provisioning lets open a separate issue if that is something we want custom docs per student.  The public sector team is super hot on that... but I feel like if people are literally typing the wrong student number in I am worried they are literally not learning anything... I guess the goal is more NOT to slow down other students with a problem student

**Comment by goetzrieger** [comment]
> @goetzrieger we had a UX person come and update our workshop page to match ansible.com
> https://ansible.github.io/workshops/

Looks a lot more professional.

> for the "per student" provisioning lets open a separate issue if that is something we want custom docs per student. 

This is something we did ages ago with some quick&dirty PHP... I really like this being brought up again. I can't count how often I had to tell students "no, your username is not student<X>" ;)

So while I understand Sean's reservations about students paying even less attention... but hey, if they spare some processing cycles with this unimportant stuff they got more left for the actual lab tasks!

So I say let's go for this. There is another option to consider, Hugo as a site-generator takes variables. So it would be enough to clone the repo, change the variable "username" (or whatever) and run Hugo to create the content locally during deployment.

**Comment by IPvSean** [comment]
OK, I am going to bring up student based documentation b/c this has been brought up to me from several angels.  I really need to spend time creating an architecture with the translation team that makes sense to do this... lets track this separately from the look/feel though of documentation.  Please open a new issue for specific look/feel, or for dynamic docs if I don't get to it before someone else does!
